enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,159 @@
+
+2. **External OAuth identity providers**
+
+For login events external OAuth providers must invoke oauth-server callback handler to finalize authorization using [code grant flow](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1).
+
+Login success will be emitted upon successful code exchange with oauth-server in the [central OAuth callback handler](https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/oauth/external/handler.go#L149-L157). 
+
+Login failure however cannot be easily detected as there is no guarantee that external oauth providers will invoke callback in the case of a login failure. Hence, only best-effort login failure events can be emitted, if the external OAuth provider provides an `error` response in the [central `HandleRequest` handler](https://github.com/openshift/oauth-server/blob/513a8bb5b6cbd5e8faacbed523eb861aa29a674b/vendor/github.com/RangelReale/osincli/authorize.go#L84).
+
+3. **External OAuth identity providers used as challengers**
+
+External oauth identity providers used as challengers use a [central location](https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/oauth/external/handler.go#L118) where login and login failures audit events will be emitted.
+
+4. **Request Header identity provider**
+
+Login events can easily be constructed by creating audit events in the [central request authentication method](https://github.com/openshift/oauth-server/blob/4e63f0f350f28171edc4ae2553c7b852e65efbc6/pkg/authenticator/request/headerrequest/requestheader.go#L34).
+
+### Logout events
+
+Logout events are not visible to oauth-server as the infrastructure in OpenShift does not have a central logout handler mechanism. Instead we have to rely on audit events against deletion of `oauthaccesstoken` resources.
+
+It is an open question how we can add information about the underlying user of the deleted `oauthaccesstoken` resource.
+
+### Event construction
+
+In all cases above a [Info](https://github.com/openshift/oauth-server/blob/961295a3151da3fff9daa5ea0ab8c0bf92a7d7e6/vendor/k8s.io/apiserver/pkg/authentication/user/user.go#L20) entity is available with user information which can be used to fill audit event data.
+
+### Integration in must-gather
+
+The existing `oc adm must-gather -- /usr/bin/gather_audit_logs` tool will also fetch oauth-server audit logs.
+
+### Open Questions
+
+An initial idea was to add annotations during `oauthaccesstoken` creation and deletion to indicate what underlying user is being logged out. However, that information is already included [in the `oauthaccesstoken` type](https://github.com/openshift/oauth-apiserver/blob/958c914e2b0b2d58f9b51715b2b76aaae774d72b/pkg/oauth/apis/oauth/types.go#L28).
```

Comment:
the information is missing during deletion though, isn't it?

Reply:
The object does include the necessary information during the in-flight deletion request, it is just not fully printed as it is done in the case of a creation event.

Reply:
so we would need to find a way to add it

Reply:
as discussed, crazy idea: add an in-process admission plugin into oauth-apiserver which intercepts deletions of `oauthaccesstoken` resources and creates a dedicated "logout" event with all necessary information. I will play around with this idea and see if that works.

Reply:
this has been resolved and the solution is way easier than assumed :)